src/components/KCodeBlock/KCodeBlock.vue

```diff
@@ -0,0 +1,829 @@
+        class="k-line-number-rows"
+      >
+        <a
+          v-for="line in matchingLineNumbers"
+          :id="`${linePrefix}-L${line}`"
+          :key="line"
+          :href="`#${linePrefix}-L${line}`"
+        >{{ line }}</a>
+      </span><code v-html="filteredCode" /></pre>
+
+      <pre
+        v-else
+        class="k-highlighted-code-block"
+        data-testid="k-highlighted-code-block"
+      ><span
+        v-if="props.showLineNumbers"
+        class="k-line-number-rows"
+      >
+        <a
+          v-for="line in totalLines"
+          :id="`${linePrefix}-L${line}`"
+          :key="line"
+          class="k-line"
+          :class="{
+            'k-line--is-match': matchingLineNumbers.includes(line),
+            'k-line--is-highlighted-match': currentLineIndex !== null && line === matchingLineNumbers[currentLineIndex],
+          }"
+          :href="`#${linePrefix}-L${line}`"
+        >{{ line }}</a>
+      </span><code v-html="props.code" /></pre>
+      <!-- eslint-enable vue/no-v-html -->
+
+      <KClipboardProvider
+        v-if="props.showCopyButton"
+        v-slot="{ copyToClipboard }"
+      >
+        <KPop placement="bottom">
+          <KButton
+            class="k-code-block-copy-button"
+            type="button"
+            appearance="outline"
+            :is-rounded="false"
+            size="small"
+            title="Copy"
+            data-testid="k-code-block-copy-button"
+            @click="copyToClipboard(props.code)"
+          >
+            <KIcon
+              icon="copy"
+              size="24"
+              title="Copy"
+              color="currentColor"
+            />
+
+            <span class="k-visually-hidden">Copy</span>
+          </KButton>
+
+          <template #content>
+            Code copied to clipboard
+          </template>
+        </KPop>
+      </KClipboardProvider>
+    </div>
+  </div>
+</template>
+
+<script lang="ts" setup>
+import { computed, nextTick, onBeforeUnmount, onMounted, ref, watch } from 'vue'
+
+import KButton from '@/components/KButton/KButton.vue'
+import KClipboardProvider from '@/components/KClipboardProvider/index'
+import KIcon from '@/components/KIcon/KIcon.vue'
+import KInput from '@/components/KInput/KInput.vue'
+import KLabel from '@/components/KLabel/KLabel.vue'
+import KPop from '@/components/KPop/KPop.vue'
+import { debounce } from '@/utils/debounce'
+
+export type CodeBlockEventData = {
+  preElement: HTMLElement
+  codeElement: HTMLElement
+  code: string
+  language: string
+  query: string
+  matchingLineNumbers: number[]
+}
+
+// Debounces the search handler which ensures that we don’t trigger several searches while the user is still typing.
+const debouncedHandleSearchInputValue = debounce(handleSearchInputValue, 150)
+
+const props = defineProps({
+  /**
+   * ID value used for form elements like the search input and its label.
+   */
+  id: {
+    type: String,
+    required: false,
+    default: 'code-block',
```

Comment:
The default here should be an empty string, and then in the template, you should set the attribute to `undefined` (so it does not render) if the prop equates to an empty string.

Otherwise, having 2 `KCodeBlock` elements on the same page with no `id` prop provided will result in invalid HTML (2 elements with the same `id` on a page).

Reply:
That’s in my opinion a worse developer experience because it prevents you from using a single instance of `KCodeBlock` in a document without specifying `props.id`. Both when not providing a default value and when providing one will it be necessary for two `KCodeBlock` components to have distinct IDs so I’d consider the approach of having a default value the one with less friction because it helps prevent developer error.

That being said, perhaps `props.id` should actually be required. While this wouldn’t allow for using one code block while omitting its ID it would always alert on a missing one when using two code blocks which currently could silently cause errors (though it’s very likely as the code block is very unlikely to participate in a form context which is the prime candidate for ID conflict related errors).

For now I’m changing `props.id` to be required.

Reply:
If you make it required and remove the default, I think that's better; we'll just need to outline in the docs that `id` must be unique (in case it isn't obvious).

Reply:
Good point; it’s not immediately obvious that a `props.id` is going to be basis for a form control `id` attribute. I’ll add a note to the props.id` section in the docs.